transports/websys-webtransport/src/connection.rs

```diff
@@ -0,0 +1,180 @@
+use libp2p_core::multihash::Multihash;
+use libp2p_core::muxing::{StreamMuxer, StreamMuxerEvent, StreamMuxerExt};
+use libp2p_core::{OutboundUpgrade, UpgradeInfo};
+use libp2p_identity::{Keypair, PeerId};
+use send_wrapper::SendWrapper;
+use std::collections::HashSet;
+use std::future::poll_fn;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+use wasm_bindgen_futures::JsFuture;
+use web_sys::{ReadableStreamDefaultReader, WebTransport, WebTransportBidirectionalStream};
+
+use crate::cached_js_promise::CachedJsPromise;
+use crate::endpoint::Endpoint;
+use crate::utils::{parse_reader_response, to_js_type};
+use crate::{Error, Stream};
+
+/// An opened WebTransport connection.
+pub struct Connection {
+    session: SendWrapper<WebTransport>,
+    create_stream_promise: CachedJsPromise,
+    incoming_stream_promise: CachedJsPromise,
+    incoming_streams_reader: SendWrapper<ReadableStreamDefaultReader>,
+    closed: bool,
+}
+
+impl Connection {
+    pub(crate) fn new(endpoint: &Endpoint) -> Result<Self, Error> {
+        let url = endpoint.url();
+
+        let session = if endpoint.certhashes.is_empty() {
+            // Endpoint has CA-signed TLS certificate
+            WebTransport::new(&url).map_err(Error::from_js_value)?
+        } else {
+            // Endpoint has self-signed TLS certificates
+            let opts = endpoint.webtransport_opts();
+            WebTransport::new_with_options(&url, &opts).map_err(Error::from_js_value)?
+        };
+
+        let incoming_streams = session.incoming_bidirectional_streams();
+        let incoming_streams_reader =
+            to_js_type::<ReadableStreamDefaultReader>(incoming_streams.get_reader())?;
+
+        Ok(Connection {
+            session: SendWrapper::new(session),
+            create_stream_promise: CachedJsPromise::new(),
+            incoming_stream_promise: CachedJsPromise::new(),
+            incoming_streams_reader: SendWrapper::new(incoming_streams_reader),
+            closed: false,
+        })
+    }
+
+    pub(crate) async fn authenticate(
+        &mut self,
+        keypair: &Keypair,
+        remote_peer: Option<PeerId>,
+        certhashes: HashSet<Multihash>,
+    ) -> Result<PeerId, Error> {
+        self.ready().await?;
+        let stream = self.create_stream().await?;
+
+        let noise = libp2p_noise::Config::new(keypair)?.with_webtransport_certhashes(certhashes);
+
+        // We do not use `upgrade::apply_outbound` function because it uses
+        // `multistream_select` protocol, which is not used by WebTransport spec.
+        let info = noise.protocol_info().next().unwrap_or_default();
+        let (peer_id, _io) = noise.upgrade_outbound(stream, info).await?;
+
+        // TODO: This should be part libp2p-noise
+        if let Some(expected_peer_id) = remote_peer {
+            if peer_id != expected_peer_id {
+                return Err(Error::UnknownRemotePeerId);
+            }
+        }
```

Comment:
Is this needed? If yes, then should I implement `libp2p_noise::Config::with_remote_peer_id` and do the check in libp2p-noise?